core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -1045,6 +1053,69 @@ public RelBuilder snapshot(RexNode period) {
     return this;
   }
 
+
+  /**
+   * Gets column mappings of the operator.
+   *
+   * @param op operator instance
+   * @return column mappings associated with this function
+   */
+  private Set<RelColumnMapping> getColumnMappings(SqlOperator op) {
+    SqlReturnTypeInference inference = op.getReturnTypeInference();
+    if (inference == null) {
+      return null;
+    }
+    if (inference instanceof TableFunctionReturnTypeInference) {
+      return ((TableFunctionReturnTypeInference) inference).getColumnMappings();
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Create a RexCall to the CURSOR function by ordinal.
+   *
+   * @param ordinal the reference to the relational input
+   * @return RexCall to CURSOR function
+   */
+  public RexNode cursor(int ordinal) {
+    return getRexBuilder().makeCall(SqlStdOperatorTable.CURSOR, getRexBuilder()
+        .makeInputRef(peek(stack.size() - ordinal - 1).getRowType(), ordinal));
+  }
+
+  /**
+   * Creates a {@link TableFunctionScan}.
+   *
+   */
+  public RelBuilder functionScan(SqlOperator operator,
+      Iterable<? extends RexNode> operands) {
+    List<RelNode> inputs = new ArrayList<>();
+    int total = stack.size();
+
+    // Get inputs from operands.
+    for (RexNode operand : operands) {
+      if (operand instanceof RexCall) {
+        if (operand.getKind() == SqlKind.CURSOR) {
+          RexNode node = ((RexCall) operand).getOperands().get(0);
+          assert node instanceof RexInputRef;
+          inputs.add(peek(total - ((RexInputRef) node).getIndex() - 1));
+        }
+      }
+    }
+
+    RexNode call = getRexBuilder()
+        .makeCall(operator, ImmutableList.copyOf(operands));
```

Comment:
You don't need to add make immutable copy. The RexCall constructor will do.

Reply:
It do not target to make immutable copy. It converts operands whose type is Iterable<? extends RexNode> to List using ImmutableList#copyOf.

Reply:
Got it. Make sense.